File: frontend/src/types/property.ts
export interface Property {
  id: string;
  landlord_id: string;
  name: string;
  address: string;
  city: string;
  district: string;
  total_units: number;
  description?: string;
  created_at: string;
  updated_at: string;
}

export interface PropertyCreate {
  name: string;
  address: string;
  city: string;
  district: string;
  description?: string;
}

export interface PropertyUpdate {
  name?: string;
  address?: string;
  city?: string;
  district?: string;
  description?: string;
}

export type UnitStatus = "vacant" | "occupied" | "maintenance";

export interface Unit {
  id: string;
  property_id: string;
  unit_number: string;
  floor?: number;
  bedrooms: number;
  bathrooms: number;
  area_sqft?: number;
  rent_amount: number;
  status: UnitStatus;
  created_at: string;
  updated_at: string;
}

export interface UnitCreate {
  unit_number: string;
  floor?: number;
  bedrooms: number;
  bathrooms: number;
  area_sqft?: number;
  rent_amount: number;
}

export interface UnitUpdate {
  unit_number?: string;
  floor?: number;
  bedrooms?: number;
  bathrooms?: number;
  area_sqft?: number;
  rent_amount?: number;
  status?: UnitStatus;
}

File: backend/app/schemas/property.py
from pydantic import BaseModel, field_validator
from typing import Optional
from uuid import UUID
from datetime import datetime


class PropertyCreate(BaseModel):
    name: str
    address: str
    city: str
    district: str
    description: Optional[str] = None


class PropertyUpdate(BaseModel):
    name: Optional[str] = None
    address: Optional[str] = None
    city: Optional[str] = None
    district: Optional[str] = None
    description: Optional[str] = None
    image_url: Optional[str] = None


class PropertyResponse(BaseModel):
    id: str
    landlord_id: str
    name: str
    address: str
    city: str
    district: str
    total_units: int
    description: Optional[str] = None
    image_url: Optional[str] = None
    created_at: datetime
    updated_at: datetime

    @field_validator("id", "landlord_id", mode="before")
    @classmethod
    def uuid_to_str(cls, v):
        return str(v)

    class Config:
        from_attributes = True